Qa filter integration from context

gaborod16 opened the pull request:
Add generic context to the BPI QA integration to leverage the URL filter of the QA system

gaborod16 reviewed (approved):
Looks solid!
Thanks for adding the defaults as well.

Note that, we need to merge the other PR first and then re-check this one @MohamedYaman

gaborod16 reviewed (approved):
Approved after a pair-programming session with @MohamedYaman

Commit:
Adapt logging levels

Commit:
Update requirements.txt

Commit:
Update BPI version

Commit:
Add helper and constants

Commit:
Add helped

MohamedYaman reviewed (approved):
I like it Picasso

Commit:
Add helpers to extract uuids from paths

Commit:
Extend functionality to check if a session exists and handle manually context injection for qa-url-filter context

Commit:
Remove unnecessary timer

Commit:
Allow the tracking of a QA response as a CAI response and track it as an additional session step

Commit:
Add autocoded endpoints

kNalj commented on ondewo_bpi/autocoded/agent_grpc_autocode.py:
```diff
@@ -38,7 +38,7 @@ class AutoAgentsServicer(AgentsServicer):
 
     [original docstring]
-    Agents are best described as Natural Language Understanding (NLU) modules
+    Agents are best described as Natural Language Understanding (NLU) modules that transform user requests into actionable data. You can include agents in your app, product, or service to determine user intent and respond to the user in a natural way.
```
PEP8 line too long

MohamedYaman commented on ondewo_bpi_qa/contants.py:
```diff
@@ -0,0 +1,7 @@
+QA_URL_FILTER_CONTEXT_NAME: str = "c-qa-url-filter"
+QA_URL_FILTER_DEFAULT_PARAM_NAME: str = 'default-filter'
```
Change the name to Base filter :smile:

MohamedYaman commented on ondewo_bpi_qa/bpi_qa_server.py:
```diff
@@ -46,19 +53,22 @@ def __init__(self) -> None:
-    def DetectIntent(self, request: DetectIntentRequest, context: grpc.ServicerContext) -> DetectIntentResponse:
+    @Timer(log_arguments=False, logger=logger_console.debug)
+    def DetectIntent(self, request: DetectIntentRequest,
+                     context: grpc.ServicerContext) -> DetectIntentResponse:
```
format bitte

kNalj replied:
If fits in one row:
  def func_name(param_1: type, param_2: type) -> return_type:


If it does not fit in one row
  def func_name(
      param_1: type,
      param_2: type,
  ) -> return_type:

NOT THE SAME AS:
  def func_name(
      param_1: type,
      param_2: type,) -> return_type:

But a question is: What if all params fit in the second row ?
  def func_name(
      param_1: type, param_2: type
  ) -> return_type:

Any thoughts ?

gaborod16 replied:
So you think @kNalj

    @Timer(log_arguments=False, logger=logger_console.debug)
    def DetectIntent(self,
            request: DetectIntentRequest,
            context: grpc.ServicerContext
    ) -> DetectIntentResponse:

is more readable than:

    @Timer(log_arguments=False, logger=logger_console.debug)
    def DetectIntent(self,
                     request: DetectIntentRequest,
                     context: grpc.ServicerContext) -> DetectIntentResponse:

?

kNalj replied:
```
@Timer(log_arguments=False, logger=logger_console.debug)
def DetectIntent(self,
                 request_with_a_super_long_name: DetectIntentRequest,
                 another_parameter_with_very_long_name: DetectIntentReqest
                 context: grpc.ServicerContext) -> DetectIntentResponse:
```

What about this case ?

MohamedYaman commented on ondewo_bpi_qa/bpi_qa_server.py:
```diff
@@ -160,12 +275,85 @@ async def send_to_cai(self, request: DetectIntentRequest,) -> Tuple[DetectIntent
+
+        try:
+            self.client.services.sessions.get_session(
+                request=GetSessionRequest(session_id=request.session, session_view=Session.View.VIEW_SPARSE)
```
Is this a good thing that when you get a session, it throws an error if there is none?
I mean I know it's the existing system behaviour but I am not a fan of that

gaborod16 replied:
I think the original flaw is the fact that Sessions are not created explicitly.
If we could guarantee that you cannot use `DetectIntent` without a created session, this logic wouldn't need to exist.

kNalj commented on ondewo_bpi_qa/bpi_qa_server.py:
```diff
@@ -35,6 +39,9 @@
     QA_PORT,
     SESSION_TIMEOUT_MINUTES,
 )
+from ondewo_bpi_qa.contants import QA_URL_FILTER_CONTEXT_NAME, QA_URL_FILTER_DEFAULT_PARAM_NAME, \
```
Is this supposed to be "conStants" ?

MohamedYaman replied:
yes typo

kNalj commented on ondewo_bpi_qa/bpi_qa_server.py:
```diff
@@ -46,19 +53,22 @@ def __init__(self) -> None:
+        response, response_name = self.handle_predictions(request=request)
 
-        if response_name == "cai_response":
+        if response_name == CAI_RESPONSE_NAME:
```
:1st_place_medal:

kNalj commented on ondewo_bpi_qa/bpi_qa_server.py:
```diff
@@ -79,18 +89,84 @@ def check_session_id(self, request: DetectIntentRequest) -> None:
-    @Timer(log_arguments=False)
-    def handle_async(self, request: DetectIntentRequest,) -> Tuple[DetectIntentResponse, str]:
+    @Timer(log_arguments=False, logger=logger_console.debug)
+    def handle_context_injection_for_qa(self, request: DetectIntentRequest, ) -> DetectIntentRequest:
```
:100:

MohamedYaman commented on ondewo_bpi_qa/bpi_qa_server.py:
```diff
@@ -79,18 +89,84 @@ def check_session_id(self, request: DetectIntentRequest) -> None:
+                continue
+
+            context_to_inject: Context = Context(
+                name=context.name,
```
Can't you just do
context_to_inject: Context = Context()
context_to_inject.CopyFrom(context)

 So it's kinda simpler than writing everything out

MohamedYaman reviewed (changes requested):
Changes bitte

kNalj commented on ondewo_bpi_qa/helper.py:
```diff
@@ -0,0 +1,48 @@
+        return path.rsplit('/', maxsplit=1)[-1]
+
+
+if __name__ == '__main__':
```
:1st_place_medal:

Commit:
Address PR comments:

 * Refactor comments on the autocoded file for agent grpc interface
 * Simplify copy of the qa context to inject
 * Rename constants for easier usage

kNalj commented on ondewo_bpi_qa/helper.py:
```diff
@@ -0,0 +1,48 @@
+        match: Optional[Match] = cls.AGENT_PATTERN.search(path)
+        if match is None:
+            raise ValueError(f'Given agent name "{path}" has invalid format. '
+                             f'Required format: "projects/<Project ID>/agent/...".')
```
Is raising an exception a desired behavior ?

gaborod16 replied:
I would say so.
The system will always give you a proper format.

If it doesn't, you want your integration system to tell you that something's wrong. :) 

Also, if the user inputs something wrong it would fail in the system anyways (nlu)

kNalj reviewed (commented):
Noice.

Commit:
Update release notes

Commit:
Fix small issues:
 - Wrong usage of library
 - Improve login when no filter is set
 - Correct code to set the default filter if the context is set but not the filter parameters
